File: src/PcIfy.Server/Constants/MediaTypes.cs
namespace PcIfy.Server.Constants;

public static class MediaTypes
{
    private static readonly Dictionary<string, string> MimeMap = new(StringComparer.OrdinalIgnoreCase)
    {
        // Video
        [".mp4"]  = "video/mp4",
        [".mkv"]  = "video/x-matroska",
        [".avi"]  = "video/x-msvideo",
        [".mov"]  = "video/quicktime",
        [".wmv"]  = "video/x-ms-wmv",
        [".flv"]  = "video/x-flv",
        [".webm"] = "video/webm",
        [".m4v"]  = "video/x-m4v",
        [".3gp"]  = "video/3gpp",
        // Image
        [".jpg"]  = "image/jpeg",
        [".jpeg"] = "image/jpeg",
        [".png"]  = "image/png",
        [".gif"]  = "image/gif",
        [".bmp"]  = "image/bmp",
        [".webp"] = "image/webp",
        [".heic"] = "image/heic",
        [".tiff"] = "image/tiff",
        // Audio
        [".mp3"]  = "audio/mpeg",
        [".flac"] = "audio/flac",
        [".aac"]  = "audio/aac",
        [".ogg"]  = "audio/ogg",
        [".wav"]  = "audio/wav",
        [".m4a"]  = "audio/mp4",
        // Document
        [".pdf"]  = "application/pdf",
        [".txt"]  = "text/plain",
        [".zip"]  = "application/zip",
    };

    public static string GetMimeType(string extension) =>
        MimeMap.TryGetValue(extension, out var mime) ? mime : "application/octet-stream";

    public static bool IsVideo(string extension) =>
        MimeMap.TryGetValue(extension, out var mime) && mime.StartsWith("video/");

    public static bool IsImage(string extension) =>
        MimeMap.TryGetValue(extension, out var mime) && mime.StartsWith("image/");

    public static bool IsAudio(string extension) =>
        MimeMap.TryGetValue(extension, out var mime) && mime.StartsWith("audio/");

    public static bool IsThumbnailable(string extension) =>
        IsVideo(extension) || IsImage(extension);
}

File: src/PcIfy.Server.Flutter/lib/core/constants/media_types.dart
abstract final class MediaTypes {
  static const _mimeMap = <String, String>{
    'mp4': 'video/mp4',
    'mkv': 'video/x-matroska',
    'avi': 'video/x-msvideo',
    'mov': 'video/quicktime',
    'wmv': 'video/x-ms-wmv',
    'flv': 'video/x-flv',
    'webm': 'video/webm',
    'm4v': 'video/x-m4v',
    '3gp': 'video/3gpp',
    'jpg': 'image/jpeg',
    'jpeg': 'image/jpeg',
    'png': 'image/png',
    'gif': 'image/gif',
    'bmp': 'image/bmp',
    'webp': 'image/webp',
    'heic': 'image/heic',
    'tiff': 'image/tiff',
    'tif': 'image/tiff',
    'mp3': 'audio/mpeg',
    'flac': 'audio/flac',
    'aac': 'audio/aac',
    'ogg': 'audio/ogg',
    'wav': 'audio/wav',
    'm4a': 'audio/mp4',
    'pdf': 'application/pdf',
    'txt': 'text/plain',
    'doc': 'application/msword',
    'docx':
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'xls': 'application/vnd.ms-excel',
    'xlsx':
        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'zip': 'application/zip',
    'rar': 'application/vnd.rar',
    '7z': 'application/x-7z-compressed',
    'tar': 'application/x-tar',
    'gz': 'application/gzip',
  };

  static const _videoExts = {
    'mp4', 'mkv', 'avi', 'mov', 'wmv', 'flv', 'webm', 'm4v', '3gp',
  };
  static const _imageExts = {
    'jpg', 'jpeg', 'png', 'gif', 'bmp', 'webp', 'heic', 'tiff', 'tif',
  };
  static const _audioExts = {'mp3', 'flac', 'aac', 'ogg', 'wav', 'm4a'};

  static String getMimeType(String extension) {
    return _mimeMap[extension.toLowerCase()] ?? 'application/octet-stream';
  }

  static bool isVideo(String extension) =>
      _videoExts.contains(extension.toLowerCase());

  static bool isImage(String extension) =>
      _imageExts.contains(extension.toLowerCase());

  static bool isAudio(String extension) =>
      _audioExts.contains(extension.toLowerCase());

  static bool isThumbnailable(String extension) =>
      isVideo(extension) || isImage(extension);
}
